File: sections/35_math_mixed_contest/problems/f_festival_synchronization/solution.py
import sys


def extended_gcd(a: int, b: int) -> tuple[int, int, int]:
    if b == 0:
        return a, 1, 0
    g, x1, y1 = extended_gcd(b, a % b)
    return g, y1, x1 - (a // b) * y1


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    q = data[0]
    output = []
    index = 1

    for _ in range(q):
        r1, m1, r2, m2 = data[index:index + 4]
        index += 4
        g, inverse, _ = extended_gcd(m1, m2)
        difference = r2 - r1
        if difference % g:
            output.append("-1")
            continue

        reduced_modulus = m2 // g
        multiplier = (difference // g * inverse) % reduced_modulus
        period = m1 // g * m2
        answer = (r1 + m1 * multiplier) % period
        output.append(str(answer))

    print("\n".join(output))


if __name__ == "__main__":
    main()

File: sections/35_math_mixed_contest/problems/f_festival_synchronization/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using int64 = long long;

int64 extended_gcd(int64 a, int64 b, int64& x, int64& y) {
    if (b == 0) {
        x = 1;
        y = 0;
        return a;
    }
    int64 x1, y1;
    int64 g = extended_gcd(b, a % b, x1, y1);
    x = y1;
    y = x1 - (a / b) * y1;
    return g;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    cin >> q;
    while (q--) {
        int64 r1, m1, r2, m2;
        cin >> r1 >> m1 >> r2 >> m2;

        int64 inverse, unused;
        int64 g = extended_gcd(m1, m2, inverse, unused);
        int64 difference = r2 - r1;
        if (difference % g != 0) {
            cout << -1 << '\n';
            continue;
        }

        int64 reduced_modulus = m2 / g;
        int64 multiplier = static_cast<int64>(
            (__int128)(difference / g) * inverse % reduced_modulus
        );
        if (multiplier < 0)
            multiplier += reduced_modulus;

        int64 period = m1 / g * m2;
        int64 answer = static_cast<int64>(
            ((__int128)m1 * multiplier + r1) % period
        );
        cout << answer << '\n';
    }
}
